features/dnsresolver/netty/src/main/java/org/opennms/netmgt/dnsresolver/netty/CaffeineCache.java

```diff
@@ -0,0 +1,138 @@
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.dnsresolver.netty;
+
+import java.util.Collection;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.TimeUnit;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Expiry;
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+
+/**
+ * Underlying cache used to back the {@link CaffeineDnsCache}.
+ *
+ * Uses the Caffeine caching library as opposed to Netty's io.netty.resolver.dns.Cache implementation so that we can:
+ *  1) Limit the cache's size (and intelligently evict entries)
+ *  2) Expose cache stats
+ *
+ * @author jwhite
+ */
+public abstract class CaffeineCache<E> {
+
+    private final Cache<String, Entries<E>> cache;
+
+    /**
+     * Returns {@code true} if this entry should replace all other entries that are already cached for the hostname.
+     */
+    protected abstract boolean shouldReplaceAll(E entry);
+
+    public CaffeineCache(long maxSize) {
+        final Caffeine<String, Entries> caffeine = Caffeine.newBuilder()
+                .expireAfter(new Expiry<String, Entries>() {
+                    @Override
+                    public long expireAfterCreate(String key, Entries value, long currentTime) {
+                        // Use the TTL of the first entry added for the key as the expiry time
+                        return TimeUnit.SECONDS.toNanos(value.ttl());
+                    }
+
+                    @Override
+                    public long expireAfterUpdate(String key, Entries value, long currentTime, long currentDuration) {
+                        // Don't expire after update
+                        return currentDuration;
+                    }
+
+                    @Override
+                    public long expireAfterRead(String key, Entries value, long currentTime, long currentDuration) {
+                        // Don't expire after read
+                        return currentDuration;
+                    }
+                })
+                .recordStats();
+        if (maxSize > 0) {
+            // Apply a size limit, if set
+            caffeine.maximumSize(maxSize);
+        }
+        cache = caffeine.build();
+    }
+
+    public void clear() {
+        cache.invalidateAll();
+    }
+
+    public void clear(String key) {
+        cache.invalidate(key);
+    }
+
+    public Collection<E> get(String key) {
+        final Entries<E> entries = cache.getIfPresent(key);
+        if (entries == null) {
+            return null;
+        }
+        return entries.get();
+    }
+
+    public void cache(String key, E entry, int ttl) {
+        cache.get(key, k -> new Entries<>(ttl)).add(entry, shouldReplaceAll(entry));
```

Comment:
If `shouldReplaceAll(...)` returns `true`, this can be simplified to a `cache.put(...)`.

Reply:
Method was revised while changing the TTL logic. Not sure this has much value in changing anymore.

Reply:
I'm fine with this, too.